### PR TITLE
MotionProfiling Controller

### PR DESCRIPTION
**Pull request by JacisNonsense**
Migrates the MP code from last season with a closer structure to this year. Strategy and bindings still need to be done.

### CODE REVIEW

File: common/src/main/cpp/Drivetrain.cpp [hunk 1]
```diff
@@ -21,6 +21,20 @@ void curtinfrc::Drivetrain::SetInverted(bool inverted) {
   }
 }
 
+double curtinfrc::Drivetrain::GetLeftDistance() {
+  auto gb = GetLeft();
+
+  assert(gb.encoder != nullptr);
+  return gb.encoder->GetEncoderRotations() * 3.14159265 * 2 * _config.wheelRadius;
```
Comment (met4000) [inline common/src/main/cpp/Drivetrain.cpp]: * 3.14159265...  🤔

File: common/src/main/cpp/MotionProfiling.cpp [hunk 2]
```diff
@@ -1,76 +1,71 @@
...
+void PathfinderController::SetOffset(double distanceL, double distanceR) {
+  _offsetL = distanceL;
+  _offsetR = distanceR;
+} 
+
+std::pair<double, double> PathfinderController::Calculate(double distL, double distR, double gyroAngle) {
+  double l = pathfinder_follow_distance(_cfg, &_followerL, _segmentsL, _trajLen, distL - _offsetL);
+  double r = pathfinder_follow_distance(_cfg, &_followerR, _segmentsR, _trajLen, distR - _offsetR);
+
+  double desiredAngle = -r2d(_followerL.heading);
```
Comment (met4000) [inline common/src/main/cpp/MotionProfiling.cpp]: Just me getting triggered a little bit (I know it doesn't actually change anything as each side is the same), but on this line:
(https://github.com/CurtinFRC/2019-DeepSpace/blob/b54ec61469c6dd2bf8754d8779c61d82bd0f3803/common/src/main/cpp/MotionProfiling.cpp#L49)
you get the generic trajectory length from the right hand side, but here you get the bearing from the left hand side...

Comment (JacisNonsense) [reply common/src/main/cpp/MotionProfiling.cpp]: The heading and file lengths are guaranteed to be identical, so it's not really a big problem

File: common/src/main/include/MotionProfiling.h [hunk 3]
```diff
@@ -1,104 +1,73 @@
...
+    double GetkD() const;
+    double GetkA() const;
+    double GetkV() const;
+    double GetkG() const;
 
-  private:
-    sensors::Encoder *_enc;
+   private:
+    std::shared_ptr<nt::NetworkTable> _table;
+    double _kP, _kI, _kD, _kA, _kV, _kG;
```
[the remark this answers is not in the document]
Comment (JacisNonsense) [reply common/src/main/include/MotionProfiling.h]: Good point, we'll have to cover it in docs

**Pull request merged**